### ISSUE TITLE
Secret Caching

### ISSUE THREAD

**Issue by Ashnam1100** [dapr/dapr#3514]
## In what area(s)?

 /area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Describe the feature

Allow the Dapr sidecar to cache the secret it retrieves from any secret store.

Use case: When using other secret retrieval tools, we've run into throttling issues because pods called into the Azure Key Vault too many times. With secret caching, the Dapr sidecar will need to retrieve the secret from AKV fewer times, still provide the pods with the secret as many times as required, and eliminate throttling problems.

Suggestions:
- Make time to live in cache customizable
- Make cache maxsize customizable
- Ensure cache is secure, especially since it would hold secrets

## Release Note

RELEASE NOTE: **ADD** secret caching.

**Comment by artursouza** [comment]
> @Ashnam1100 Would [dapr/components-contrib#1037](https://github.com/dapr/components-contrib/issues/1037) be an acceptable alternative? The app would need to manage the local cache as a separate store, but it might be a better first step as it focuses on the security of the local cache component.

I think this is about an optimization to avoid calling the backing service multiple times. Issue dapr/components-contrib#1037 is about local development.

**Comment by artursouza** [comment]
Considerations for working on this issue:
- In-memory cache must be encrypted.
- Cache TTL must be configurable via component metadata.
- Allow getSecret API to invalidate the cache entry (force retrieve)

**Comment by greenie-msft** [comment]
Doc PR here: https://github.com/dapr/docs/pull/2019

**Comment by 0xakihiko** [comment]
This looks mostly finished in https://github.com/dapr/dapr/pull/3941 - any reason it was removed from the v1.7 milestones?

**Comment by mukundansundar** [comment]
@Taction are you working on this issue and if so do you have an ETA for this?

**Comment by Taction** [comment]
@mukundansundar Yes, I am doing this in #3941, and I'll need some time to clear the conflict. It could be finished by this weekend or next weekend.

**Comment by peter-alegeus** [comment]
+1 for this issue. Key Vault throttling is a problem on the horizon for multiple projects I am and will be working on, and it would make things easier if this could be handled in the secrets component rather than within each individual service that needs caching.
